From 07a31c130647d479f7ea11e0731150ecf47a5d10 Mon Sep 17 00:00:00 2001
From: Erico Nunes <nunes.erico@gmail.com>
Date: Tue, 16 Nov 2021 16:44:35 +0100
Subject: [PATCH 1/2] cogl/framebuffer: Remove requirement to discard color
 buffer

cogl_framebuffer_discard_buffers required that the color buffer is
passed, although users might want to discard e.g. just depth and/or
stencil buffers.

Signed-off-by: Erico Nunes <nunes.erico@gmail.com>
---
 cogl/cogl/cogl-framebuffer.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/cogl/cogl/cogl-framebuffer.c b/cogl/cogl/cogl-framebuffer.c
index 525b24293f..4930ed51dc 100644
--- a/cogl/cogl/cogl-framebuffer.c
+++ b/cogl/cogl/cogl-framebuffer.c
@@ -1695,8 +1695,6 @@ cogl_framebuffer_discard_buffers (CoglFramebuffer *framebuffer,
   CoglFramebufferPrivate *priv =
     cogl_framebuffer_get_instance_private (framebuffer);
 
-  g_return_if_fail (buffers & COGL_BUFFER_BIT_COLOR);
-
   cogl_framebuffer_driver_discard_buffers (priv->driver, buffers);
 }
 
-- 
GitLab


From f4001683fbb6bfaef3ae3e1077107e69ed09bdb0 Mon Sep 17 00:00:00 2001
From: Erico Nunes <nunes.erico@gmail.com>
Date: Tue, 16 Nov 2021 16:44:39 +0100
Subject: [PATCH 2/2] cogl/onscreen: Discard depth/stencil before swapping
 buffers

Marking the the depth/stencil as discarded before swapping buffers for
the screen signals the GPU that we don't need to keep them around for
the future.
This helps performance by reducing memory bandwidth usage in some GPUs
which may optimize to not write those buffers back to memory at all
after rendering, when they would just be cleared right after that
anyway.

Signed-off-by: Erico Nunes <nunes.erico@gmail.com>
---
 cogl/cogl/cogl-onscreen.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/cogl/cogl/cogl-onscreen.c b/cogl/cogl/cogl-onscreen.c
index ff9e1749a7..9d95862311 100644
--- a/cogl/cogl/cogl-onscreen.c
+++ b/cogl/cogl/cogl-onscreen.c
@@ -334,6 +334,10 @@ cogl_onscreen_swap_buffers_with_damage (CoglOnscreen *onscreen,
   if (G_UNLIKELY (COGL_DEBUG_ENABLED (COGL_DEBUG_SYNC_FRAME)))
     cogl_framebuffer_finish (framebuffer);
 
+  cogl_framebuffer_discard_buffers (framebuffer,
+                                    COGL_BUFFER_BIT_DEPTH |
+                                    COGL_BUFFER_BIT_STENCIL);
+
   klass->swap_buffers_with_damage (onscreen,
                                    rectangles,
                                    n_rectangles,
@@ -395,6 +399,10 @@ cogl_onscreen_swap_region (CoglOnscreen *onscreen,
      COGL_WINSYS_FEATURE_SWAP_REGION */
   g_return_if_fail (klass->swap_region);
 
+  cogl_framebuffer_discard_buffers (framebuffer,
+                                    COGL_BUFFER_BIT_DEPTH |
+                                    COGL_BUFFER_BIT_STENCIL);
+
   klass->swap_region (onscreen,
                       rectangles,
                       n_rectangles,
-- 
GitLab

